docs/proposals/2022-07-13-appset-progressive-rollout-strategy.md

```diff
@@ -0,0 +1,169 @@
+---
+title: ApplicationSet Progressive Rollout Strategy
+authors:
+  - "@wmgroot"
+  - "@cnmcavoy"
+sponsors:
+  - indeed.com
+reviewers:
+  - "@alexmt"
+  - TBD
+approvers:
+  - "@alexmt"
+  - TBD
+
+creation-date: 2022-07-13
+last-updated: 2022-07-13
+---
+
+# ApplicationSet Progressive Rollout Strategy
+
+## Summary
+
+Enhance the ArgoCD ApplicationSet resource to embed a rollout strategy for a progressive application resource update after the ApplicationSet spec or Application templates are modified.
+Further discussion and interest has been communicated here: https://github.com/argoproj/argo-cd/issues/9437
+
+## Motivation
+
+As cluster operators, we would like to make changes to ApplicationSets which may target multiple environments, pre-defined staging areas, or other configurations, and have these changes rolled out in a declarative, defined manner rather than all at once as ApplicationSets currently behave. A progressive ApplicationSet rollout would prevent mistakes in configuration from having a larger blast radius than intended and give cluster operators a chance to verify and have confidence in their changes.
+
+### Goals
+
+Users are able to make a single change to ApplicationSet that is updated across the generated Applications in a controlled manner. When this enhancement is enabled, Applications are updated in a declaractive order, instead of simultaneously.
+
+### Non-Goals
+
+Handling controlled rollouts for changes to a helm chart or raw manifests referenced by the Applications managed by the ApplicationSet. We understand this would be valuable, but we would like to implement the rollout implementation handling only changes to the ApplicationSet initially.
+
+## Proposal
+
+This is where we get down to details of what the proposal is about.
+
+### Use cases
+
+Add a list of detailed use cases this enhancement intends to take care of.
+
+#### Use case 1:
+As a user, I would like to declaratively control the rollout order of ApplicationSet changes to its generated Application resources.
+
+We propose adding a `RollingUpdate` strategy spec (taking inspiration from other controllers).
+
+The rolling update strategy deterministically chooses applications to update one by one, proceeding each step only if the previous application syncs completed successfully, and respecting the maxUpdate value.
```

Comment:
> one by one

Also, in the Goals section:
> Applications are updated in a declarative order

I would like to also add the use case of being able to define sets of env/clusters/etc to be done simultaneously.  For example

I might update a canary cluster, qa-1, in QA updated and then want the rest of the clusters in QA updated all in one go.  Once that is synced, then continue to a prod canary, prod-1, and then update the rest of the clusters in Prod in bulk sets. Point is I would like to provide the ability to avoid updating one cluster at a time in order to speed parts of the deployment process up.

The above set notation might look something like

1. [qa-1]
2. [qa-2,qa-3]
3. [prod-1]
4. [prod-2,prod-3]
5. [prod-4,prod-5]

Reply:
Oh, it looks like that is what MaxUpdate does.